### ISSUE TITLE
stack deploy fails due to missing service that is not missing

### ISSUE THREAD

**Issue by pcornelissen** [kontena/kontena#1958]
I recreated my staging grid and now I get an error when I try to install my stack:

```
# kontena stack install -n staging
 [fail] Creating stack staging
 [error] {"services"=>"Service create failed for service 'restaurants-host': Service staging/contactme-host does not exist"}
```

The kontena yaml contains the requested service of course:

```
services:
....
  contactme-host:
    extends:
      file: common-kontena.yaml
      service: java-service
    image: "nexus-dockerkulinariweb.orchit-dev.de/kulinariweb/contactme:${STAGE}_LATEST"
    depends_on:
      - eureka
      - confighub
      - rabbitmq
      - redis
      - dbinit
    links:
      - mariadb/node:mariadb
...
  restaurants-host:
    extends:
      file: common-kontena.yaml
      service: java-service
    image: "nexus-dockerkulinariweb.orchit-dev.de/kulinariweb/restaurants:${STAGE}_LATEST"
    depends_on:
      - eureka
      - rabbitmq
      - confighub
      - contactme-host
    links:
      - mariadb/node:mariadb
....
```
The services are created and in the initialized state. A kontena stack deploy afterwards seems to work. (still running)

**Comment by SpComb** [comment]
With a bit of logging, it looks like the server `Stack::SortHelper` is getting confused by the `sort_services(GridService)` vs `sort_services({services: [ ... ]}[:services])` variations in the `links` structure:

When called from `Stacks::Create` with the hash variant:

```
sort_services <- ["redis-before", "redis", "redis-after"]
sort_services a=redis links=[] <=> b=redis-after links=[{"name"=>"redis", "alias"=>"redis"}]
sort_services a=redis-before links=[{"name"=>"redis", "alias"=>"redis"}] <=> b=redis links=[]
sort_services a=redis-before links=[{"name"=>"redis", "alias"=>"redis"}] <=> b=redis-after links=[{"name"=>"redis", "alias"=>"redis"}]
sort_services -> ["redis", "redis-before", "redis-after"]
```

When called from `StackRemoveWorker` with the `GridService` model variant:

```
sort_services <- ["redis-before", "redis", "redis-after"]
sort_services a=redis links=[] <=> b=redis-after links=[{:name=>"redis"}]
sort_services a=redis-before links=[{:name=>"redis"}] <=> b=redis links=[]
sort_services a=redis-before links=[{:name=>"redis"}] <=> b=redis-after links=[{:name=>"redis"}]
sort_services -> ["redis", "redis-before", "redis-after"]
```

The `links` is `{ :name => ... }` vs `{ "name" => ... }`, and the implementation only works with the former, as used in the stack remove case...

**Comment by SpComb** [comment]
I think what's happening here is that one of your linked stack services has some invalid attributes (such as a `depends_on: ` referring to itself) which causes it to fail to be created. However, the server continues creating the remaining services, and then overwrites the actual error with this confusing error from the service that was linking to the service that actually failed to create.

**Comment by SpComb** [comment]
Assuming that the `staging/contactme-host` service failed to create for some reason, and then that cascaded to the `restaurants-host` service failing to create, then #1976 should fix the error handling, and also return the actual error that caused the `staging/contactme-host` service create to fail.

Assuming that the underlying error was caused by some service in the stack linking to itself, then #1976 should also fix that to return an `Linked service '...' refers to self` error.

**Comment by pcornelissen** [comment]
The contactme-host has been initialized and deploy of the whole stack works fine. I just have to start the deploy manually

**Comment by SpComb** [comment]
Odd, the `Service create failed for service 'restaurants-host': ...` should indicate that the `restaurants-host` service should not exist.

Can you provide the full stack file somehow? You must have some complex linking structure that is confusing the server if this is happening every time with Kontena 1.1.

**Comment by SpComb** [comment]
It seems the server's stack service sorting algorithm does get confused by appropriately complex stack service linking schemes... presumably when a service `bar` links to a `foo` later in the stack file, which then in turn links to a service `asdf` that has more links than either of `foo` or `bar`... The associated `kontena stack install` and `kontena stack rm` depend on the exact ordering of the services in the stackfile.

Repro case:

```yaml
stack: test/test
services:
  bar:
    image: redis
    depends_on:
      - foo

  foo:
    image: redis
    depends_on:
      - asdf

  asdf:
    image: redis
    depends_on:
      - asdf1
      - asdf2
      - asdf3
  asdf1:
    image: redis
  asdf2:
    image: redis
  asdf3:
    image: redis
```

With the error handling improvements in #1976:
```
 [fail] Creating stack test      
 [error] {"services"=>{"bar"=>{"links"=>"Service test/foo does not exist"}}}
```

In this case, the `test/bar` service is missing from `kontena service ls`. Changing the order of the services in the stack file may allow them to be created correctly, but may also results in associated `kontena stack rm` hangs as in #1836:

```
127.0.0.1 - - [15/Mar/2017:11:22:03 +0200] "DELETE /v1/stacks/development/test HTTP/1.1" 200 2 0.0869
E, [2017-03-15T11:22:03.742477 #32302] ERROR -- StackRemoveWorker: Cannot remove service development/test/asdf: {"service"=>"Cannot delete service that is linked to another service (foo)"}
E, [2017-03-15T11:22:05.795839 #32302] ERROR -- StackRemoveWorker: Cannot remove service development/test/asdf3: {"service"=>"Cannot delete service that is linked to another service (asdf)"}
E, [2017-03-15T11:22:05.813132 #32302] ERROR -- StackRemoveWorker: Cannot remove service development/test/asdf2: {"service"=>"Cannot delete service that is linked to another service (asdf)"}
E, [2017-03-15T11:22:05.832220 #32302] ERROR -- StackRemoveWorker: Cannot remove service development/test/asdf1: {"service"=>"Cannot delete service that is linked to another service (asdf)"}
```

**Comment by pcornelissen** [comment]
These are the created services. It looks like all are created. (They are running because I started a stack deploy afterwards)
```
pcornelissen@Geldgrab:~/Documents/projects/kulinariweb/contactme (develop *>)$ kontena service ls
NAME                                                         INSTANCES  STATEFUL STATE      EXPOSED PORTS
⊛ staging/frontend-host                                      2 / 2      no       running
⊛ staging/geodata-host                                       2 / 2      no       running
⊛ staging/lunchdeals-host                                    2 / 2      no       running
⊛ staging/contactme-host                                     2 / 2      no       running
⊛ staging/adminproxy-host                                    2 / 2      no       running
⊛ staging/adminusers-host                                    2 / 2      no       running
⊛ staging/restaurants-host                                   2 / 2      no       running
⊛ staging/contracts-host                                     2 / 2      no       running
⊛ staging/registration-host                                  2 / 2      no       running
⊛ staging/dashboard-host                                     2 / 2      no       running
⊛ staging/fullmenu-host                                      2 / 2      no       running
⊛ staging/emailverificator-host                              2 / 2      no       running
⊛ staging/assets-host                                        2 / 2      no       running
⊛ staging/frontendusers-host                                 2 / 2      no       running
⊛ staging/subscriptions-host                                 2 / 2      no       running
⊝ staging/turbine-host                                       1 / 1      no       running
⊛ staging/confighub                                          2 / 2      no       running
⊛ staging/eureka                                             1 / 1      no       running
⊝ staging/dbinit                                             1 / 1      no       running
⊛ staging/statics                                            2 / 2      no       running
⊝ staging/adminuiv2                                          2 / 2      no       running
⊛ staging/rabbitmq                                           1 / 1      no       running
⊛ staging/freegeoip                                          1 / 1      yes      running
⊝ staging/redis                                              1 / 1      yes      running
⊝ staging/loadbalancer                                       2 / 1      no       running    0.0.0.0:80->80/tcp, 0.0.0.0:443->443/tcp
⊝ mariadb/dbbackup                                           1 / 1      no       running
⊛ mariadb/node                                               3 / 3      no       running
⊛ mariadb/seed                                               1 / 1      yes      running
⊝ vpn/server                                                 1 / 1      yes      running    0.0.0.0:1194->1194/udp
```

**Comment by pcornelissen** [comment]
BTW: This behaviour was introduced somewhere in 1.1.3 or 1.1.4

**Comment by SpComb** [comment]
To clarify, after this happens:

```
# kontena stack install -n staging
 [fail] Creating stack staging
 [error] {"services"=>"Service create failed for service 'restaurants-host': Service staging/contactme-host does not exist"}
```

Does `kontena service ls` show that the `staging/restaurants-host` service is missing? That is, before you run any `kontena stack deploy staging`, which will indeed create any missing services.

I'm guessing that the bug here is #1980, and thus the workaround would be to add more direct `depends_on: ...` links to any indirectly linked services: if service A depends on service B, and service B depends on service C, then without the #1980 fix, service A must also depend on service C.

**Comment by pcornelissen** [comment]
OK, I just removed the stack and added it again with 1.1.5.rc2:
```
$ kontena stack install -n staging
 [fail] Creating stack staging
 [error] {"services"=>"Service create failed for service 'restaurants-host': Service staging/contactme-host does not exist"}
$ kontena service ls
NAME                                                         INSTANCES  STATEFUL STATE      EXPOSED PORTS
⊗ staging/frontend-host                                      0 / 2      no       initialized
⊗ staging/lunchdeals-host                                    0 / 2      no       initialized
⊗ staging/geodata-host                                       0 / 2      no       initialized
⊗ staging/contactme-host                                     0 / 2      no       initialized
⊗ staging/adminproxy-host                                    0 / 2      no       initialized
⊗ staging/assets-host                                        0 / 2      no       initialized
⊗ staging/contracts-host                                     0 / 2      no       initialized
⊗ staging/dashboard-host                                     0 / 2      no       initialized
⊗ staging/frontendusers-host                                 0 / 2      no       initialized
⊗ staging/emailverificator-host                              0 / 2      no       initialized
⊗ staging/adminusers-host                                    0 / 2      no       initialized
⊗ staging/fullmenu-host                                      0 / 2      no       initialized
⊗ staging/subscriptions-host                                 0 / 2      no       initialized
⊝ staging/turbine-host                                       0 / 1      no       initialized
⊗ staging/confighub                                          0 / 2      no       initialized
⊗ staging/eureka                                             0 / 1      no       initialized
⊝ staging/dbinit                                             0 / 1      no       initialized
⊗ staging/statics                                            0 / 2      no       initialized
⊝ staging/adminuiv2                                          0 / 2      no       initialized
⊗ staging/rabbitmq                                           0 / 1      no       initialized
⊗ staging/freegeoip                                          0 / 1      yes      initialized
⊝ staging/redis                                              0 / 1      yes      initialized
⊝ staging/loadbalancer                                       0 / 1      no       initialized 0.0.0.0:80->80/tcp, 0.0.0.0:443->443/tcp
```
You are right, the restaurants-host service is missing! And I made the mistake of trying to remove the "half-baked" stack and the services changed from initialized to stopped, but no further action was taken. I tried to call rm multiple times afterwards with no effect at all.
Now I'll have to wait until the deploy is finished (rm during deploy gets ignored too) to retry to remove the stack. (the trick was to restart the master after(!) the stack deploy added the missing services... otherwise all rm calls had no effect)

Had another try where kontena was complaining, now I have to wait for the deployment... again

**Comment by SpComb** [comment]
This should be fixed in 1.3.0 once released.
